### PR TITLE
feat: Add more powerful evaluation methods

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

Mostly motivated by #442

Add a more powerful abstraction that allows running multiple evaluation handlers within one pipeline. Also, support different techniques for scheduling and sampling data for evaluations.

# Changes

Mostly production logic changes are in `/supervisor/internal/eval` and `pipeline_executor.py`

Add `EvalTrigger`s:
- `AfterTrainingTrigger`: Run evaluation after every trigger/training
- `StaticEvalTrigger`: Run evaluation at pre-configured timestamps
- `PeriodicEvalTrigger`: Run evaluation in a pre-configured schedule

Add / adapt `EvalStrategy`
- `IntervalEvalStrategy`: from the point of evaluation add some offset in both directions defining the interval of evaluation
- `MatrixEvalStrategy`: same behavior as before (tumbling window)

The rest of the changes come from adjusted configs (moved to subdir), test rewrites, and documentation.

### CODE REVIEW

File: modyn/tests/supervisor/internal/eval_strategies/test_matrix_eval_strategy.py [hunk 1]
```diff
@@ -1,37 +0,0 @@
-from modyn.supervisor.internal.eval_strategies import MatrixEvalStrategy
```
Comment (robinholzi) [inline modyn/tests/supervisor/internal/eval_strategies/test_matrix_eval_strategy.py:1]: moved to `internal/eval/strategies/test_interval_eval_strategy`

File: modyn/tests/supervisor/internal/eval/strategies/test_matrix_eval_strategy.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from modyn.config.schema.pipeline.evaluation import MatrixEvalStrategyConfig
```
Comment (robinholzi) [inline modyn/tests/supervisor/internal/eval/strategies/test_matrix_eval_strategy.py:1]: not new, only moved into a subdir

File: modyn/supervisor/internal/eval_strategies/offset_eval_strategy.py [hunk 3]
```diff
@@ -1,49 +0,0 @@
-from typing import Iterable, Optional
-
-from modyn.supervisor.internal.eval_strategies.abstract_eval_strategy import AbstractEvalStrategy
-from modyn.utils import convert_timestr_to_seconds
-
-
-class OffsetEvalStrategy(AbstractEvalStrategy):
```
Comment (robinholzi) [inline modyn/supervisor/internal/eval_strategies/offset_eval_strategy.py:7]: `IntervalEvalStrategy` enables a superset of features of `OffsetEvalStrategy`, therefore i dropped it. If needed, we can also keep this though.

File: modyn/config/schema/pipeline/pipeline.py [hunk 4]
```diff
@@ -456,129 +425,6 @@ class DataDriftTriggerConfig(BaseModel):
     Union[TimeTriggerConfig, DataAmountTriggerConfig, DataDriftTriggerConfig], Field(discriminator="id")
 ]
 
-
-# ---------------------------------------------------- EVALUATION ---------------------------------------------------- #
```
Comment (robinholzi) [inline modyn/config/schema/pipeline/pipeline.py:460]: moved into `schema/pipeline/evaluation.py`

File: modyn/config/schema/pipeline/evaluation.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
+"""
```
Comment (robinholzi) [inline modyn/config/schema/pipeline/evaluation.py:1]: Taken from `schema/pipeline.py` and adjusted the config according the the changes in evaluation logic.

File: modyn/config/schema/pipeline/data.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+from typing import List, Optional
```
Comment (robinholzi) [inline modyn/config/schema/pipeline/data.py:1]: taken form `schema/pipeline.py`

**Comment by robinholzi** [comment]
@MaxiBoether let me know if we should do an in-person review for shorter feedback cycles. As this is again a rather big PR, I don't expect you to review this immediately.

**Comment by MaxiBoether** [comment]
> @MaxiBoether let me know if we should do an in-person review for shorter feedback cycles. As this is again a rather big PR, I don't expect you to review this immediately.

I'll probably do a first pass of this on Friday or the weekend, just to get an understanding, and then get back to your offer :) I'll also got through the documentation PR again so we get that merged

**Comment by robinholzi** [comment]
@MaxiBoether It's probably better if you don't review this just yet. I'm currently fixing the issues of the evaluator not being able to handle more than a certain number of evaluation requests (missing memory release). For that, I'll rearrange the changes from this PR a bit. It probably makes sense to review this only when those changes are ready.

**Comment by robinholzi** [comment]
Implemented in #478, #479, #480, #481

**Pull request closed without merging**